qmcpy/integrand/AmericanOption.py

```diff
@@ -0,0 +1,115 @@
+        interest_rate=0.05, n=4096, t_final=1, call_put='put', multilevel_dims=None, _dim_frac=0, observations=52):
+        self.parameters = ['volatility', 'call_put', 'start_price', 'strike_price', 'interest_rate']
+        # handle single vs multilevel
+        self.multilevel_dims = multilevel_dims
+        self.call_put=call_put
+        self.observations=observations #number of time steps
+        self.strike_price=strike_price
+        if self.multilevel_dims is not None: # multi-level problem
+            self.dim_fracs = array(
+                [0]+ [float(self.multilevel_dims[i])/float(self.multilevel_dims[i-1]) 
+                for i in range(1,len(self.multilevel_dims))],
+                dtype=float)
+            self.max_level = len(self.multilevel_dims)-1
+            self.leveltype = 'fixed-multi'
+            self.parent = True
+            self.parameters += ['multilevel_dims']
+            self.sampler=Sobol(self.multilevel_dims),
+        else: # single level problem
+            self.dim_fracs = _dim_frac
+            self.leveltype = 'single'
+            self.parent = False
+            self.parameters += ['dim_frac']
+            self.sampler=sampler=Sobol(self.observations)
+        self.t_final = t_final
+        self.n=n # Number of samples
+        self.start_price=start_price
+        self.interest_rate = interest_rate
+        self.t=np.linspace(1/self.observations,self.t_final, self.observations)
+        self.true_measure = BrownianMotion(self.sampler,self.t_final) #Browninan motion
+        self.volatility = volatility
+        super(AmericanOption,self).__init__(dimension_indv=1,dimension_comb=1,parallel=False)    
+        
+    
+    def _get_discounted_payoffs(self):
+        self.stock_values =self.start_price*np.exp((self.interest_rate-0.5*self.volatility**2)*self.t+self.volatility*self.true_measure.gen_samples(self.n))
+        in_money_values=np.zeros(self.n,dtype=bool)
+        y = np.zeros(self.n) #output for linear regression
+        exercise_time=np.ones(self.n)
+        exercise_time=self.t_final*exercise_time
+        max_function = np.vectorize(lambda x: max(x, 0))
+        payoff = max_function(self.strike_price-self.stock_values) #Payoffs at each time (or values assuming we do not exercise at any time)
+        payoff = payoff.astype(np.float64)
+        payoff *=np.exp(-self.interest_rate*self.t)
+        values=payoff[:,self.observations-1] #Option values after comparing payoffs with linear regression
+        #print("values",values)
+        for j in range(self.observations-1,0,-1):
+            in_money_values=payoff[:,j]>0
+            x=self.stock_values[in_money_values,j-1]
+            y=values[in_money_values]
+            A = np.vstack([x, np.ones(len(x))]).T
+            m, c = np.linalg.lstsq(A, y, rcond=None)[0] #use multilinear regression
+            hold_lr=m*self.stock_values[in_money_values,j-1]+c #lr of holding value one more time step
+            exercise=in_money_values
```

Comment:
I want to note that on line 64 which is `exercise=in_money_values` could have potential problems later on. This statement does not make a copy the list but instead has `exercise` point to the same list that `in_money_values` also points to. At the moment it seems that the code works as it should due to pure luck, but this could break at a later time. However, if this was intended, then disregard this comment.

Reply:
For some reason GitHub highlighted the code above the line I've mentioned as well, even though I highlighted the code that was underneath it because it's also related. Not sure why GitHub messed that up.

Reply:
Maybe you don't need the `exercise` variable and can just use `in_money_values` instead. Otherwise use `numpy.copy`

Reply:
Just to reduce the risk of something breaking, I went ahead and used `numpy.copy` so that it would mostly stay the same.